File: kotlin/src/test/kotlin/com/solana/mpp/client/SessionConsumerTest.kt
package com.solana.mpp.client

import com.solana.mpp.crypto.MemorySigner
import com.solana.mpp.crypto.PublicKey

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

/**
 * SessionConsumer metered delivery flow. Mirrors the Rust spine
 * (client/session_consumer.rs) and the Go session_consumer_test.go:
 * the local watermark advances only after the transport accepts the commit,
 * directives are validated against the active session, and a failed commit
 * leaves the watermark untouched so a retry does not drift.
 */
class SessionConsumerTest {
    private val channel = PublicKey(ByteArray(32) { 9 })
    private val signer = MemorySigner.fromSeed(ByteArray(32) { 3 })

    private fun directive(amount: String, sessionId: String = channel.toBase58()): MeteringDirective =
        MeteringDirective(
            deliveryId = "d1",
            sessionId = sessionId,
            amount = amount,
            currency = "USDC",
            sequence = 1,
            expiresAt = DEFAULT_SESSION_EXPIRES_AT,
        )

    private class RecordingTransport(private val status: CommitStatus = CommitStatus.COMMITTED) : CommitTransport {
        var lastPayload: CommitPayload? = null
        var calls = 0
        override fun commit(directive: MeteringDirective, payload: CommitPayload): CommitReceipt {
            calls += 1
            lastPayload = payload
            return CommitReceipt(
                deliveryId = directive.deliveryId,
                sessionId = directive.sessionId,
                amount = directive.amount,
                cumulative = payload.voucher.data.cumulative,
                status = status,
            )
        }
    }

    @Test
    fun commitDirectiveSignsAdvancesAndReturnsReceipt() {
        val session = ActiveSession.create(channel, signer)
        val transport = RecordingTransport()
        val consumer = SessionConsumer(session, transport)

        val receipt = consumer.commitDirective(directive("125"))
        assertEquals(CommitStatus.COMMITTED, receipt.status)
        assertEquals("125", receipt.cumulative)
        assertEquals(125L, session.cumulative)
        assertEquals("d1", transport.lastPayload?.deliveryId)
    }

    @Test
    fun acceptAndAckSettlesDelivery() {
        val session = ActiveSession.create(channel, signer)
        val consumer = SessionConsumer(session, RecordingTransport())
        val delivery = consumer.accept(directive("200"))
        assertEquals("d1", delivery.metering().deliveryId)
        val receipt = delivery.commit()
        assertEquals("200", receipt.cumulative)
        assertEquals(200L, session.cumulative)
    }

    @Test
    fun mismatchedSessionRejected() {
        val session = ActiveSession.create(channel, signer)
        val consumer = SessionConsumer(session, RecordingTransport())
        assertFailsWith<com.solana.mpp.protocol.MppException> {
            consumer.commitDirective(directive("125", sessionId = "other-session"))
        }
    }

    @Test
    fun zeroAmountRejected() {
        val session = ActiveSession.create(channel, signer)
        val consumer = SessionConsumer(session, RecordingTransport())
        assertFailsWith<com.solana.mpp.protocol.MppException> {
            consumer.commitDirective(directive("0"))
        }
    }

    @Test
    fun failedCommitLeavesWatermarkUntouched() {
        val session = ActiveSession.create(channel, signer)
        val failing = object : CommitTransport {
            override fun commit(directive: MeteringDirective, payload: CommitPayload): CommitReceipt {
                throw com.solana.mpp.protocol.MppException.InvalidTransaction("transport down")
            }
        }
        val consumer = SessionConsumer(session, failing)
        assertFailsWith<com.solana.mpp.protocol.MppException> {
            consumer.commitDirective(directive("125"))
        }
        // Watermark unchanged so the same directive can be retried.
        assertEquals(0L, session.cumulative)

        // A subsequent successful commit for the same directive amount works.
        val ok = SessionConsumer(session, RecordingTransport())
        val receipt = ok.commitDirective(directive("125"))
        assertEquals("125", receipt.cumulative)
        assertEquals(125L, session.cumulative)
    }

    @Test
    fun replayedStatusPropagates() {
        val session = ActiveSession.create(channel, signer)
        val consumer = SessionConsumer(session, RecordingTransport(CommitStatus.REPLAYED))
        val receipt = consumer.commitDirective(directive("125"))
        assertEquals(CommitStatus.REPLAYED, receipt.status)
    }

    @Test
    fun sessionAccessor() {
        val session = ActiveSession.create(channel, signer)
        val consumer = SessionConsumer(session, RecordingTransport())
        assertTrue(consumer.session() === session)
    }
}

File: go/protocols/mpp/client/session_consumer_test.go
package client

import (
	"context"
	"errors"
	"testing"

	solana "github.com/gagliardetto/solana-go"

	"github.com/solana-foundation/pay-kit/go/protocols/mpp/intents"
)

type recordingTransport struct {
	commits []intents.CommitPayload
	fail    bool
}

func (r *recordingTransport) Commit(_ context.Context, directive intents.MeteringDirective, payload intents.CommitPayload) (intents.CommitReceipt, error) {
	if r.fail {
		return intents.CommitReceipt{}, errors.New("commit failed")
	}
	r.commits = append(r.commits, payload)
	return intents.CommitReceipt{
		DeliveryID: directive.DeliveryID,
		SessionID:  directive.SessionID,
		Amount:     directive.Amount,
		Cumulative: payload.Voucher.Data.Cumulative,
		Status:     intents.CommitStatusCommitted,
	}, nil
}

func directive(sessionID string, amount uint64) intents.MeteringDirective {
	return intents.MeteringDirective{
		DeliveryID: "d1",
		SessionID:  sessionID,
		Amount:     itoa(amount),
		Currency:   "USDC",
		Sequence:   1,
		ExpiresAt:  DefaultVoucherExpiresAt,
	}
}

func itoa(v uint64) string {
	if v == 0 {
		return "0"
	}
	var b []byte
	for v > 0 {
		b = append([]byte{byte('0' + v%10)}, b...)
		v /= 10
	}
	return string(b)
}

func TestConsumerAckCommitsAndAdvances(t *testing.T) {
	session := NewActiveSession(solana.NewWallet().PublicKey(), seededSigner(7))
	transport := &recordingTransport{}
	consumer := NewSessionConsumer(session, transport)

	delivery, err := consumer.Accept(directive(session.ChannelIDStr(), 250))
	if err != nil {
		t.Fatal(err)
	}
	receipt, err := delivery.Ack(context.Background())
	if err != nil {
		t.Fatal(err)
	}
	if receipt.Cumulative != "250" {
		t.Fatalf("receipt cumulative = %s", receipt.Cumulative)
	}
	if consumer.Session().Cumulative() != 250 {
		t.Fatalf("local watermark = %d, want 250", consumer.Session().Cumulative())
	}
	if len(transport.commits) != 1 {
		t.Fatalf("expected 1 commit, got %d", len(transport.commits))
	}
}

func TestConsumerRejectsWrongSession(t *testing.T) {
	session := NewActiveSession(solana.NewWallet().PublicKey(), seededSigner(7))
	consumer := NewSessionConsumer(session, &recordingTransport{})
	if _, err := consumer.Accept(directive("other-session", 1)); err == nil {
		t.Fatalf("expected wrong-session rejection")
	}
}

func TestConsumerRejectsZeroAmount(t *testing.T) {
	session := NewActiveSession(solana.NewWallet().PublicKey(), seededSigner(7))
	consumer := NewSessionConsumer(session, &recordingTransport{})
	if _, err := consumer.CommitDirective(context.Background(), directive(session.ChannelIDStr(), 0)); err == nil {
		t.Fatalf("expected zero-amount rejection")
	}
}

func TestConsumerFailedCommitDoesNotAdvance(t *testing.T) {
	session := NewActiveSession(solana.NewWallet().PublicKey(), seededSigner(7))
	transport := &recordingTransport{fail: true}
	consumer := NewSessionConsumer(session, transport)
	if _, err := consumer.CommitDirective(context.Background(), directive(session.ChannelIDStr(), 250)); err == nil {
		t.Fatalf("expected commit failure")
	}
	if consumer.Session().Cumulative() != 0 {
		t.Fatalf("watermark should not advance on failed commit")
	}
}
